src/storage/src/render/sources.rs

```diff
@@ -602,6 +612,23 @@ where
     }
 }
 
+fn handle_missing_keys(key_arity: Option<usize>) -> Box<dyn Fn(KV) -> KV> {
+    let null_key_columns = Row::pack_slice(&vec![Datum::Null; key_arity.unwrap_or(0)]);
```

Comment:
shouldn't this be happening in `flatten_results_prepend_key`? I think instead we should change the api of `raise_key_value_errors` to:
```
fn raise_key_value_errors(KV { key, val }: KV) -> Option<Result<(Option<Row>, Row, Diff), DecodeError>> {
```
and change `flatten_results_prepend_keys` to pack the key slice of `nulls`

Reply:
additionally, we should make a `RowPacker` to avoid some extra allocations here, but that can come after my other comments!

Reply:
would a RowPacker be more efficient than the currently `null_key_columns.clone()`? not sure how it'd be used in this case

Reply:
oh yeah the new `raise_key_value_errors` is nice to work with, that simplified the change in `flatten_results_prepend_keys` quite a bit